data_steward/analytics/table_metrics/weekly_scripts/erroneous_dates.py

```diff
@@ -0,0 +1,583 @@
+
+FROM
+
+  (SELECT
+  DISTINCT
+  mpo.src_hpo_id, COUNT(*) as num_rows
+  FROM
+  `{DATASET}.unioned_ehr_procedure_occurrence` po
+  JOIN
+  `{DATASET}._mapping_procedure_occurrence` mpo
+  ON
+  po.procedure_occurrence_id = mpo.procedure_occurrence_id
+  GROUP BY 1
+  ORDER BY num_rows DESC) total
+
+LEFT JOIN
+
+  (SELECT
+  DISTINCT
+  mpo.src_hpo_id, COUNT(*) as num_bad_rows
+  FROM
+  `{DATASET}.unioned_ehr_procedure_occurrence` po
+  JOIN
+  `{DATASET}._mapping_procedure_occurrence` mpo
+  ON
+  po.procedure_occurrence_id = mpo.procedure_occurrence_id
+  
+  WHERE
+  po.procedure_datetime < CAST('1980-01-01 00:00:00' AS TIMESTAMP)
+  OR
+  po.procedure_date < CAST('1980-01-01' as DATE)
+  
+  GROUP BY 1
+  ORDER BY num_bad_rows DESC) bad_date
+  
+ON
+bad_date.src_hpo_id = total.src_hpo_id
+
+GROUP BY 1, 2 --, 3, 4
+ORDER BY procedure_occurrence DESC
+""".format(DATASET=DATASET)
+
+procedure_occurrence_df = pd.io.gbq.read_gbq(procedure_query, dialect ='standard')
+
+procedure_occurrence_df
+
+# ## Visit Occurrence
+
+visit_query = """
+SELECT
+total.src_hpo_id,
+-- IFNULL(bad_date.num_bad_rows, 0) AS num_bad_rows, 
+-- IFNULL(total.num_rows, 0) AS num_rows,
+ROUND(IFNULL(bad_date.num_bad_rows, 0) / IFNULL(total.num_rows, 0) * 100 , 2) as visit_occurrence
+
+FROM
+
+  (SELECT
+  DISTINCT
+  mvo.src_hpo_id, COUNT(*) as num_rows
+  FROM
+  `{DATASET}.unioned_ehr_visit_occurrence` vo
+  JOIN
+  `{DATASET}._mapping_visit_occurrence` mvo
+  ON
+  vo.visit_occurrence_id = mvo.visit_occurrence_id
+  GROUP BY 1
+  ORDER BY num_rows DESC) total
+
+LEFT JOIN
+
+  (SELECT
+  DISTINCT
+  mvo.src_hpo_id, COUNT(*) as num_bad_rows
+  FROM
+  `{DATASET}.unioned_ehr_visit_occurrence` vo
+  JOIN
+  `{DATASET}._mapping_visit_occurrence` mvo
+  ON
+  vo.visit_occurrence_id = mvo.visit_occurrence_id
+  
+  WHERE
+  vo.visit_start_datetime < CAST('1980-01-01 00:00:00' AS TIMESTAMP)
+  OR
+  vo.visit_start_date < CAST('1980-01-01' as DATE)
+  
+  OR
+  
+  vo.visit_end_datetime < CAST('1980-01-01 00:00:00' AS TIMESTAMP)
+  OR
+  vo.visit_end_date < CAST('1980-01-01' as DATE)
+  
+  GROUP BY 1
+  ORDER BY num_bad_rows DESC) bad_date
+  
+ON
+bad_date.src_hpo_id = total.src_hpo_id
+
+GROUP BY 1, 2 --, 3, 4
+ORDER BY visit_occurrence DESC
+""".format(DATASET=DATASET)
+
+visit_occurrence_df = pd.io.gbq.read_gbq(visit_query, dialect ='standard')
+
+visit_occurrence_df
+
+# # Drug Exposure
+
+drug_query = """
+SELECT
+total.src_hpo_id,
+-- IFNULL(bad_date.num_bad_rows, 0) AS num_bad_rows, 
+-- IFNULL(total.num_rows, 0) AS num_rows,
+ROUND(IFNULL(bad_date.num_bad_rows, 0) / IFNULL(total.num_rows, 0) * 100 , 2) as drug_exposure
+
+FROM
+
+  (SELECT
+  DISTINCT
+  mde.src_hpo_id, COUNT(*) as num_rows
+  FROM
+  `{DATASET}.unioned_ehr_drug_exposure` de
+  JOIN
+  `{DATASET}._mapping_drug_exposure` mde
+  ON
+  de.drug_exposure_id = mde.drug_exposure_id
+  GROUP BY 1
+  ORDER BY num_rows DESC) total
+
```

Comment:
These look similar for `visit_occurrence` above. Look into Jinja for templating queries, then you can work off of an existing template and substitute visit_occurrence, drug_exposure etc. to generate the desired queries. An example of this is in the notebook `data_steward/analytics/cdr_ops/pop_retract.py`.

Reply:
@nishanthpp93 

Point to note: I intentionally refrain from creating 'general' queries and instead separate them out by table because it enables sites to more easily copy and paste our queries directly (and therefore decrease the amount of emails directed to the DRC). I see, however, how the shorter queries makes for a more efficient/legible notebook as a whole.

Do you think it would be worthwhile to 'consolidate' the queries even if doing so may ultimately confuse the HPOs?

Reply:
I see. I would then restrict using general queries for your own use. If a site uses it, you can keep it as-is, or maybe use f-strings to make them more readable.